### PR TITLE
feat: add "Debugging Gno Programs" post

### PR DESCRIPTION
**Pull request by mvertes**
This PR introduces the blog post "Debugging Gno Programs". It is written by Marc Vertes, on 26 June 2024.

### CODE REVIEW

File: posts/2024-06-28-gno-debugger/README.md
Comment (leohhhn) [inline posts/2024-06-28-gno-debugger/README.md:6]: Yaml doesn't allow @

```suggestion
authors: [mvertes]
```

File: posts/2024-06-28-gno-debugger/README.md
Comment (leohhhn) [inline posts/2024-06-28-gno-debugger/README.md:20]: Let's use the following forms to refer to our products:
- gno.land
- Gno (capitalized)
- GnoVM

This style was agreed upon internally

```suggestion
Having a good debugger is important. But the Gno language is almost Go, and gno.land itself is entirely written in Go. Could I just use the existing Go tools, i.e.  the [delve] debugger, to take control and debug my Gno programs?
```

File: posts/2024-06-28-gno-debugger/README.md
Comment (leohhhn) [inline posts/2024-06-28-gno-debugger/README.md:22]:
```suggestion
You cannot debug your *Gno* program this way because doing so would entail debugging the Gno virtual machine rather than your own program. The relevant state information would be opaque and would need to be reversed and reconstructed from internal Gno virtual machine data structures.
```

File: posts/2024-06-28-gno-debugger/README.md
Comment (leohhhn) [inline posts/2024-06-28-gno-debugger/README.md:28]:
```suggestion
The Gno debugger is fully integrated in the [gno](https://docs.gno.land/gno-tooling/cli/gno-tooling-gno) single binary, which is the tool required to build, test and run gno programs locally.
```

File: posts/2024-06-28-gno-debugger/README.md
Comment (leohhhn) [inline posts/2024-06-28-gno-debugger/README.md:30]:
```suggestion
There is no need to install a specific tool. You just have to install the `gno` tool itself with:
```

File: posts/2024-06-28-gno-debugger/README.md
Comment (leohhhn) [inline posts/2024-06-28-gno-debugger/README.md:38]:
```suggestion
We are now ready to play with Gno programs. Let's consider a simple classic example as a target program, the computation of [Fibonacci numbers]:
```

Also, I suggest actually linking to the pages, since I see that the footer has the links. WDYT?

File: posts/2024-06-28-gno-debugger/README.md
Comment (leohhhn) [inline posts/2024-06-28-gno-debugger/README.md:4]:
```suggestion
slug: gno-debugger
```

File: posts/2024-06-28-gno-debugger/README.md
Comment (leohhhn) [inline posts/2024-06-28-gno-debugger/README.md:10]:
```suggestion
In this article, we introduce the new Gno debugger feature and show how it can be used to better understand Gno programs and help fixing bugs.
```

**Review by leohhhn — APPROVED** [review]
Thanks for this 🙏 

Left a few minor comments, and would like a review from another core team member before merging.

File: posts/2024-06-28-gno-debugger/README.md
Comment (deelawn) [inline posts/2024-06-28-gno-debugger/README.md:10]:
```suggestion
In this article, we introduce the new Gno debugger feature and show how it can be used to better understand Gno programs and help to fix bugs.
```

File: posts/2024-06-28-gno-debugger/README.md
Comment (deelawn) [inline posts/2024-06-28-gno-debugger/README.md:28]:
```suggestion
The Gno debugger is fully integrated in the [gno](https://docs.gno.land/gno-tooling/cli/gno-tooling-gno) binary, which is the tool required to build, test, and run Gno programs locally.
```
Capitalize Gno for consistency.

File: posts/2024-06-28-gno-debugger/README.md
Comment (deelawn) [inline posts/2024-06-28-gno-debugger/README.md:61]:
```suggestion
When you start a program in debug mode, you are greeted by a prompt allowing you to interact with it via the terminal:
```

File: posts/2024-06-28-gno-debugger/README.md
Comment (deelawn) [inline posts/2024-06-28-gno-debugger/README.md:93]:
```suggestion
If you have already used a debugger before, like [gdb] or [lldb] for C/C++ programs, or [delve] for Go programs, the Gno debugger should look familiar; the commands are similar in their syntax and usage.
```

File: posts/2024-06-28-gno-debugger/README.md
Comment (deelawn) [inline posts/2024-06-28-gno-debugger/README.md:95]:
```suggestion
The commands can be classified in the following categories:
```

File: posts/2024-06-28-gno-debugger/README.md
Comment (deelawn) [inline posts/2024-06-28-gno-debugger/README.md:141]:
```suggestion
Note that we have used the short alias of commands: `b` for `break` and `c` for `continue`. We only need to specify the line number when setting the break point here, due to it being in the same file. Setting break points in other files requires specifying the full file path and line number.
```
This reads a bit weird. Does this communicate the same idea?

File: posts/2024-06-28-gno-debugger/README.md
Comment (deelawn) [inline posts/2024-06-28-gno-debugger/README.md:143]:
```suggestion
We can now examine the call stack which indicates the successive nested function calls up to the current location:
```

File: posts/2024-06-28-gno-debugger/README.md
Comment (deelawn) [inline posts/2024-06-28-gno-debugger/README.md:157]:
```suggestion
We see a call stack of depth 4, with call frames (local function contexts) numbered from 0 to 3, 0 being the current call level (the deepest). This information is crucial, especially when debugging recursive functions like `fib`. We know that the caller and its caller were both `fib`.
```

File: posts/2024-06-28-gno-debugger/README.md
Comment (deelawn) [inline posts/2024-06-28-gno-debugger/README.md:159]:
```suggestion
Now we want to examine the value of the local parameter `n`, for each call level:
```

File: posts/2024-06-28-gno-debugger/README.md
Comment (deelawn) [inline posts/2024-06-28-gno-debugger/README.md:197]:
```suggestion
The `up` and `down` stack navigation commands enable the debugger to display the value of local function variables and parameters for the whole call chain.
```

File: posts/2024-06-28-gno-debugger/README.md
Comment (deelawn) [inline posts/2024-06-28-gno-debugger/README.md:199]:
```suggestion
In this example, the `n` variable is simply an integer, but the `print` command is also able to handle more complex expressions to uncover the content of arbitrary maps, struct, arrays, etc using the same syntax as Go. For example, `print a.b[n]` will print as expected, with `a` being a value of type:
```
I know what this is trying to say, but it is difficult to write 😅

File: posts/2024-06-28-gno-debugger/README.md
Comment (deelawn) [inline posts/2024-06-28-gno-debugger/README.md:205]:
```suggestion
For security reasons, the `print` command will only evaluate expressions with no side effects on the virtual machine state. For example, it is not possible to perform an arithmetic operation like `print a + 2`, or to call a function like `printf f(6)`.
```

File: posts/2024-06-28-gno-debugger/README.md
Comment (deelawn) [inline posts/2024-06-28-gno-debugger/README.md:15]:
```suggestion
>
> -- <cite>Brian Kerninghan, "The Elements of Programming Style"</cite>
```
Try previewing the markdown like this

File: posts/2024-06-28-gno-debugger/README.md
Comment (deelawn) [inline posts/2024-06-28-gno-debugger/README.md:18]:
```suggestion
>
> -- <cite>Anonymous</cite>
```

File: posts/2024-06-28-gno-debugger/README.md
Comment (deelawn) [inline posts/2024-06-28-gno-debugger/README.md:32]: I'm not sure if this works on when rendering on the platform we publish on, but it doesn't work on my local preview. It does work if the equals signs are removed. Maybe remove any special symbols outside of the syntax name in all of the code blocks? I see `=` and `!` throughout the document.

**Review by deelawn — APPROVED** [review]
Content looks good. All comments are either style or grammar suggestions.

File: posts/2024-06-28-gno-debugger/README.md
Comment (mvertes) [reply posts/2024-06-28-gno-debugger/README.md]: The trailing `=` and `!` at end of markdown code delimiters were inserted by the `hackmd.io` editor, which I used to edit this post. I removed them.

**Comment by mvertes** [comment]
Thank you @deelawn for your review!

File: posts/2024-06-28-gno-debugger/README.md
Comment (zivkovicmilos) [inline posts/2024-06-28-gno-debugger/README.md:100]: Typo here, `browsing code`

File: posts/2024-06-28-gno-debugger/README.md
Comment (zivkovicmilos) [inline posts/2024-06-28-gno-debugger/README.md:101]: Typo here, `navigating`

File: posts/2024-06-28-gno-debugger/README.md
Comment (zivkovicmilos) [inline posts/2024-06-28-gno-debugger/README.md:102]: Typo here, `quitting`

**Review by zivkovicmilos — APPROVED** [review]
This should also be part of the official documentation, great read 💯

File: posts/2024-06-28-gno-debugger/README.md
Comment (leohhhn) [reply posts/2024-06-28-gno-debugger/README.md:101]:
```suggestion
- navigating the stack: `up`, `down`,
```

File: posts/2024-06-28-gno-debugger/README.md
Comment (leohhhn) [reply posts/2024-06-28-gno-debugger/README.md:102]:
```suggestion
- quitting the debugger: `detach`, `exit`.
```

**Pull request merged**
